### ISSUE TITLE
Handle providing a lambda function when chaining methods in AskScript

### ISSUE THREAD

**Issue by czerwinskilukasz1** [xFAANG/askql#43]
```
ask {
	const double = fun(s:string):string {
    s:concat(s)
	}

  text :call(fun(s:string): string {
    s:double // same as s:double()
  })
}
```
returns the value of `text` variable instead of `text` + `text`.

**Comment by mhagmajer** [comment]
Please note that `call` resource expects function as its first argument. Following arguments are the arguments for function. Much like https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/call.

If you wish to use it in chaining you would need another resource that accepts the call argument as its first argument. For example: `apply(arg, fun)`. This is a feature request.

**Comment by mhagmajer** [comment]
Please consider changing the title to be more specific. We already have "lambda function" in AskScript.

### call(fun, arg1, arg2, ..., argN)
```
fun - function to call
arg1, ..., argN - arguments
returns - result of calling function fun with provided arguments
```

#### Example 1:

```
ask {
  const f = fun { 5 }
  call(f)
}
```

returns `5`

#### Example 2 (subject to [an ongoing issue](https://github.com/xFAANG/askql/issues/46)):

```
ask {
  call(fun { 5 })
}
```

returns `5`

**Comment by mhagmajer** [comment]
`call`, `apply`, `bind` are already known to JavaScript developers. I suggest we port these directly. They are not that bad. Of course, we do not have `this` keyword so all of these don't need their first argument here.

For the one you proposing for chained operations, maybe w can name it `useFor`?

```
2:useFor(fun(a) { a * 2 }) // return 4
```

**Comment by czerwinskilukasz1** [comment]
Fine, let's do it.

**Comment by czerwinskilukasz1** [comment]
As for porting `apply` and `bind` directly from Javascript, let's create a separate issue (or issues) for them and do them later that v.1.1 as I think they are not critical features.

**Comment by czerwinskilukasz1** [comment]
Does the current `call` implementation reflect what you mean by Javascript-`call`-just-without-`this`-keyword?

**Comment by mhagmajer** [comment]
> Does the current `call` implementation reflect what you mean by Javascript-`call`-just-without-`this`-keyword?

Yes, it should (see https://github.com/xFAANG/askql/issues/46)

**Comment by czerwinskilukasz1** [comment]
FYI I got back to working on this one.

**Comment by czerwinskilukasz1** [comment]
@mhagmajer , would you rather see useFor():
1) as a regular function on AskScript and AskJsx levels (meaning it is called via `call` resource and only then a `useFor` resource is used) or rather
2) it should rather be translated directly to a `useFor` resource before reaching AskVM?

The advantage of 1) is easier implementation, however checks that the lambda is used correctly (e.g. there is only 1 arg and it's a function) won't be performed until the script is executed.

By the way, I wonder where you want to check types of variables and validate function arguments in general.

**Comment by mhagmajer** [comment]
I would opt for #1 since `callFor` can be implemented using call (has ongoing issues). I'll answer about other questions later today

**Comment by czerwinskilukasz1** [comment]
OK, I'll do so.

**Comment by czerwinskilukasz1** [comment]
By the way, `useFor` or `callFor`?

**Comment by czerwinskilukasz1** [comment]
Blocked on #46 .

**Comment by mhagmajer** [comment]
Just to give you a small pointer, this now works in the CLI:

```
🦄 .editor
// Entering editor mode (^D to finish, ^C to cancel)
let myCall = fun(a, b) { a(b) }
myCall(toUpperCase, 'marcin')

string ask(let('myCall',fun(let('a',get('$0')),let('b',get('$1')),call(get('a'),get('b')))),call(get('myCall'),get('toUpperCase'),'marcin'))
'MARCIN'
```
